google/gax/__init__.py

```diff
@@ -482,3 +489,160 @@ def __next__(self):
+            client (google.cloud.gapic.longrunning.operations_api.OperationsApi):
+                the client used to manage the long-running operation with an API
+                service.
+            result_type (type): the class type to be unpacked from the result.
+            metadata_type (type): the class type to be unpacked from the
+                metadata.
+            call_options (google.gax.CallOptions, optional): the call options
+                that are used when reloading the operation.
+        """
+        self._first_operation = operation
+        self._last_operation = operation
+        self._client = client
+        self._result_type = result_type
+        self._metadata_type = metadata_type
+        self._call_options = call_options
+        self._cancelled = False
+
+    def cancel(self):
+        """If last Operation's value of `done` is true, returns false;
+           otherwise, issues OperationsApi.cancel_operation and returns true.
+        """
+        if not self._last_operation.done:
+            self._client.cancel_operation(self._last_operation.name)
+            self._cancelled = True
+
+        return self._cancelled
+
+    def result(self, timeout=None):
+        """Enters polling loop on OperationsApi.get_operation, and once Operation.done
+           is true, then returns Operation.response if successful or throws
+           ResultError if not successful.
+
+           This method will wait up to timeout seconds. If the call hasn't
+           completed in timeout seconds, then a concurrent.futures.TimeoutError
+           will be raised. timeout can be an int or float. If timeout is not
+           specified or None, there is no limit to the wait time.
+        """
+        if not self._poll(timeout).HasField('response'):
+            raise ResultError.from_operation_error(self._last_operation.error)
+
+        return _from_any(self._result_type, self._last_operation.response)
+
+    def exception(self, timeout=None):
+        """Similar to result(), except returns the exception if any."""
+        if self._poll(timeout).HasField('error'):
+            return self._last_operation.error
+
+        return None
+
+    def cancelled(self):
+        """Return True if the call was successfully cancelled."""
+        return self._cancelled
+
+    def done(self):
+        """Issues OperationsApi.get_operation and returns value of Operation.done."""
+        return self._get_operation().done
+
+    def add_done_callback(self, done_clbk):
+        """Enters a polling loop on OperationsApi.get_operation, and once the
+           operation is done or cancelled, calls the function with this OperationFuture.
+        """
+        def _execute_clbk(self, clbk):
+            self._poll()
+            try:
+                clbk(self)
+            except Exception as ex:  # pylint: disable=broad-except
+                _LOG.exception(ex)
+
+        threading.Thread(target=_execute_clbk, args=(self, done_clbk)).start()
```

Comment:
I think this will cause many more api calls than are necessary since each thread will do it's own polling. A solution would be to have the logic for this method be as follows:

1. If the operation is done, execute the callback and return
2. Add the callback to a list of callbacks
3. If not started, start polling thread that once polling is completed call all callbacks in the list.